main/modes/mode_dance.c

```diff
@@ -0,0 +1,1247 @@
+{
+    ledDances[danceState->danceIdx].func(elapsedUs, ledDances[danceState->danceIdx].arg, danceState->resetDance);
+
+    if (danceState->updateScreen)
+    {
+        danceRedrawScreen();
+    }
+
+    danceState->resetDance = false;
+}
+
+void danceButtonCb(buttonEvt_t* evt)
+{
+    if (evt->down)
+    {
+        if (evt->state & UP)
+        {
+            incLedBrightness();
+            danceState->updateScreen = true;
+        }
+
+        if (evt->state & DOWN)
+        {
+            decLedBrightness();
+            danceState->updateScreen = true;
+        }
+
+        if (evt->state & LEFT)
+        {
+            selectPrevDance();
+        }
+
+        if (evt->state & RIGHT)
+        {
+            selectNextDance();
+        }
+
+        if (evt->state & BTN_A)
+        {
+            danceState->blankScreen = !danceState->blankScreen;
+            danceState->updateScreen = true;
+        }
+
+        if (evt->state & BTN_B)
+        {
+
+        }
+    }
+}
+
+/**
+ * @brief Blanks and redraws the entire screen, and clears danceState->updateScreen
+ */
+void danceRedrawScreen()
+{
+    danceState->updateScreen = false;
+    danceState->disp->clearPx();
+
+    if (!danceState->blankScreen)
+    {
+        uint16_t width = textWidth(&(danceState->infoFont), ledDances[danceState->danceIdx].name);
+        drawText(danceState->disp, &(danceState->infoFont), c555, ledDances[danceState->danceIdx].name, (danceState->disp->w - width) / 2, 60);
+
+        char brightnessText[14];
+
+        snprintf(brightnessText, sizeof(brightnessText), "Brightness: %d", getLedBrightness());
+        width = textWidth(&(danceState->infoFont), brightnessText);
+        drawText(danceState->disp, &(danceState->infoFont), c555, brightnessText, (danceState->disp->w - width) / 2, 10);
+    }
+}
+
+/** @return The number of different tances
+ */
+uint8_t getNumDances(void)
+{
+    return (sizeof(ledDances) / sizeof(ledDances[0]));
+}
+
+/**
+ * @brief Get the Dance Name
+ *
+ * @param idx  The index of the dance
+ * @return the dance name
+ */
+char* getDanceName(uint8_t idx)
+{
+    return ledDances[idx].name;
+}
+
+/** This is called to clear specific dance variables
+ */
+void danceClearVars(uint8_t idx)
+{
+    // Reset the specific dance
+    ledDances[idx].func(0, ledDances[idx].arg, true);
+}
+
+/** Get a random number from a range.
+ *
+ * This isn't true-random, unless bound is a power of 2. But it's close enough?
+ * The problem is that rand() returns a number between [0, 2^64), and the
+ * size of the range may not be even divisible by bound
+ *
+ * For what it's worth, this is what Arduino's random() does. It lies!
+ *
+ * @param bound An upper bound of the random range to return
+ * @return A number in the range [0,bound), which does not include bound
+ */
+uint32_t danceRand(uint32_t bound)
+{
+    if(bound == 0)
+    {
+        return 0;
+    }
+    return rand() % bound;
+}
+
+/** This animation is set to be called every 100 ms
```

Comment:
Comment is misleading, but that may not be your fault

Reply:
Ah, I actually didn't see these and I didn't realize different modes wanted different call intervals. I thought they accounted for that by knowing the elapsed time already. Does that mean some of the animations are going to be slower/faster than they're supposed to now?

Reply:
Actually it looks like the old firmware just used one timer and always calls it as fast as possible (1ms). So I think this isn't an issue, I'll update the comments.

Reply:
No, iirc the animations on the hexagon had a parameter to call the function at some fixed rate, but that was scrapped for tracking how much time had elapsed. Don't worry about it.